cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -404,59 +392,29 @@ func (c *Client) addMember(name string, peerAddrs string, isLearner bool) ([]Mem
 				lastError = err
 				return false, nil
 			}
-			respMembers = resp.Members
 			return true, nil
 		}
 
-		resp, err = cli.MemberAdd(ctx, []string{peerAddrs})
+		_, err = cli.MemberAdd(ctx, []string{peerAddrs})
 		if err == nil {
-			respMembers = resp.Members
 			return true, nil
 		}
 
-		// If the error indicates that the peer already exists, exit early. In this situation, resp is nil, so
-		// call out to MemberList to fetch all the members before returning.
+		// If the error indicates that the peer already exists, return true.
 		if errors.Is(err, rpctypes.ErrPeerURLExist) {
 			klog.V(5).Info("The peer URL for the added etcd member already exists. Fetching the existing etcd members")
-			var listResp *clientv3.MemberListResponse
-			listResp, err = cli.MemberList(ctx)
-			if err == nil {
-				respMembers = listResp.Members
-				return true, nil
-			}
+			return true, nil
 		}
 
 		klog.V(5).Infof("Failed to add etcd member: %v", err)
 		lastError = err
 		return false, nil
 	})
 	if err != nil {
-		return nil, lastError
-	}
-
-	// Returns the updated list of etcd members
-	ret := []Member{}
-	for _, m := range respMembers {
-		// If the peer address matches, this is the member we are adding.
-		// Use the name we passed to the function.
-		if peerAddrs == m.PeerURLs[0] {
-			ret = append(ret, Member{Name: name, PeerURL: peerAddrs})
-			continue
-		}
-		// Otherwise, we are processing other existing etcd members returned by AddMembers.
-		memberName := m.Name
-		// In some cases during concurrent join, some members can end up without a name.
-		// Use the member ID as name for those.
-		if len(memberName) == 0 {
-			memberName = strconv.FormatUint(m.ID, 16)
-		}
-		ret = append(ret, Member{Name: memberName, PeerURL: m.PeerURLs[0]})
+		return lastError
```

Comment:
this is a significant refactor.
the previous code handled some edge cases where members had missing names or were added twice, during join retries. we saw some major complains about that in the past.

i don't think the new code handles that.

Reply:
You're totally right. I re-added the code to handle members that has no name yet.